vm_phys.c

```diff
@@ -0,0 +1,251 @@
+                      pg_it->phys_addr < end)) {
+                split_page(seg, pg_it);
+                /* Go to start of list because it's corrupted */
+                pg_it = TAILQ_FIRST(&seg->free_queues[i]);
+                continue;
+            }
+            pg_it = TAILQ_NEXT(pg_it, freeq);
+        }
+    }
+}
+
+void vm_phys_reserve(vm_paddr_t start, vm_paddr_t end) {
+    
+    struct vm_phys_seg *seg_it;
+    TAILQ_FOREACH(seg_it, &seglist, segq) {
+        if(seg_it->start <= start && seg_it->end >= end) {
+            vm_phys_reserve_from_seg(seg_it, start, end);
+            return;
+        }
+    }
+
+}
+
+static struct vm_page * vm_phys_alloc_from_seg(struct vm_phys_seg *seg, size_t order) {
+    struct vm_page *res = NULL;
+    size_t ord = order; 
+    /* lowest non-empty order higer than order */
+    while(TAILQ_EMPTY(&seg->free_queues[ord]) && ord < VM_NFREEORDER)
+        ord++;
+
+    if(ord == VM_NFREEORDER)
+        return NULL;
+
+    while(TAILQ_EMPTY(&seg->free_queues[order])) {
+        split_page(seg, TAILQ_FIRST(&seg->free_queues[ord]));
+        ord--;
+    }
+
+    res = TAILQ_FIRST(&seg->free_queues[order]);
+    TAILQ_REMOVE(&seg->free_queues[order], res, freeq);
+
+    res->size = ORDER_SIZE(res->order)*PAGE_SIZE;
+    return res;
+}
+
+struct vm_page * vm_phys_alloc(size_t order) {
+
+    struct vm_phys_seg *seg_it;
+    TAILQ_FOREACH(seg_it, &seglist, segq) {
+        struct vm_page *res;
+        res = vm_phys_alloc_from_seg(seg_it, order);
+        if(res != NULL) return res;
+    }
+
+    panic("vm_phys_alloc: cannot allocate page of given order");
+    return NULL;
+}
+
+static void vm_phys_free_from_seg(struct vm_phys_seg *seg, struct vm_page *page) {
+
+    /* check if page is reserved */
+    if((page->flags & VM_RESERVED) != 0)
+        panic("vm_phys_free: trying to free reserved page: %p\n", page->phys_addr);
+
+    /* check if page was free'd before */
+    struct vm_page *it;
+    TAILQ_FOREACH(it, &seg->free_queues[page->order], freeq) {
+        if(it == page)
+            panic("vm_phys_free: page is already free: %p\n", page->phys_addr);
+    }
+
+    TAILQ_INSERT_HEAD(&seg->free_queues[page->order], page, freeq);
+    struct vm_page *buddy = find_buddy(seg, page);
+    while(buddy != NULL) {
+        TAILQ_REMOVE(&seg->free_queues[buddy->order], buddy, freeq);
+        TAILQ_REMOVE(&seg->free_queues[page->order], page, freeq);
+        page = coalesce(page, buddy);
+
+        TAILQ_INSERT_HEAD(&seg->free_queues[page->order], page, freeq);
+        buddy = find_buddy(seg, page);
+    }
+}
+
+void vm_phys_free(struct vm_page *page) {
+    struct vm_phys_seg *seg_it;
+    TAILQ_FOREACH(seg_it, &seglist, segq) {
+        if(page->phys_addr >= seg_it->start && 
+            page->phys_addr + page->size <= seg_it->end  )
+        {
+            vm_phys_free_from_seg(seg_it, page);
+            return;
+        }
+    }
+    panic("vm_phys_free: page out of range\n");
```

Comment:
Good! But it'd be even nicer if you dumped segment list `(start, end)` and page address under question :)